### DISCUSSION TITLE
How to use zustand/vanilla on Node/Express & Workers?

### DISCUSSION THREAD

**Discussion by HarikalarKutusu** [pmndrs/zustand#1756] [Q&A]
I've been using zustand with React for a couple of years now and with my new project which uses microservices on server side I wanted to try the "vanilla flavor" (Typescript) and failed to understand how I can do it. My scenario in the microservice where I try is:

- I have a Bree scheduler which uses Workers to do some jobs
- The main thread initializes all and runs node/express server and implements the job chaining logic
- I want to keep logqer (winston), scheduler (Bree) and a job Queue (p-queue) in the store, so that these can be used everywhere.

So:
- Will zustand/vanilla work in such a scenario?
- I failed to implement the useStore from example snippets, so I cannot get info to the main thread from `store.ts` due to the following line complaining that the third parameter should not be there: `return useStore(store, selector!, equals);`
- Is there a complete example you can share/point to (for Typescript if possible)?

Thank you in advance...

**Comment by dbritto-dev** [comment]
@HarikalarKutusu let me answer your questions.

- Yes
- Yeah, that's expected. The `useStore` hook doesn't work outside of react app
- Sure, I'll back with an example

Note: I'm talking about zustand v4

**Reply by HarikalarKutusu** [reply]
> @HarikalarKutusu let me answer your questions.
> 
> * Yes
> * Yeah, that's expected. The `useStore` hook doesn't work outside of react app
> * Sure, I'll back with an example
> 
> Note: I'm talking about zustand v4

I think the answer to the first question should be:
- No / Not as it is

We can only pass the store objects to workers via a SharedBuffer (as parameters in Bree case).

Workers are different programs than the main thread and can share data only this way. I never used workers with larger data structures such as these, only passed simple parameters such as file names etc.

On the other hand, that will not solve my problem, because they will be a critical resource and we can only pass a snapshot at that time of spawning them. When we are talking with multiple workers using the same resource, it will collapse.

I need to find another mechanism - I think.

**Reply by HarikalarKutusu** [reply]
I think this can be solved like this:
- I can use the store only in the main thread
- Pass only the parameters required for the worker, never use store objects in there.
- Implement a job-dispatcher in the main thread. When a job finishes, it will decide to spawn other jobs if they are required.
- The vanilla store might only be useful to share data between components in the main thread like a singleton.

**Comment by HarikalarKutusu** [comment]
Thank you for the answers @dbritto-dev, to be clear: As the "hook" concept is a general one, not explicit to React, I thought the following code I've got from the examples was the implementation/imitation of it for the vanilla version:
```js
function useStore(): StoreType;
function useStore<T>(selector: (state: StoreType) => T, equals?: (a: T, b: T) => boolean): T;
function useStore<T>(selector?: (state: StoreType) => T, equals?: (a: T, b: T) => boolean) {
  return useStore(store, selector!, equals);
}
```

**Reply by dbritto-dev** [reply]
@HarikalarKutusu Maybe I'm wrong but so far I know there's no `useStore` on zustand (vanilla). Here's the code for the react one https://github.com/pmndrs/zustand/blob/main/src/react.ts#L25-L49. BTW, `hook` in this context is not a general concept is a concept tied to react.

**Reply by dbritto-dev** [reply]
Here's an example of zustand (vanilla)

```ts
import { createStore } from 'zustand/vanilla'

type Store = {
  count: number
}

const store = createStore<Store>((setState, getState) => ({
  count: 0,
  inc: () => setState((state) => ({ count: state.count + 1 })),
}))

console.log(store.getState().count)

store.getState().inc()

console.log(store.getState().count)
```

**Reply by HarikalarKutusu** [reply]
I had similar constructs, I was not importing useStore from zustand.
One important question thou... In react we had a problem with render loop, where the setMyVar does not set myVar immediately, so we had to use useEffect etc. These do not exist in vanilla... So think of this:
```js
// store.ts
import { createStore } from "zustand/vanilla";
import { Logger } from "winston";

export type StoreType = {
  // Scheduler Logger
  sLogger: Logger | null;
  setSLogger: (l: Logger) => void;
};

const store = createStore<StoreType>()((set) => ({
  // Scheduler Logger
  sLogger: null,
  setSLogger: (l) => set((state) => ({ ...state, sLogger: l })),
}

export { store };
```

```js
// index.ts
import { store } from "./stores/store";

const { getState, setState} = store;
const { sLogger, setSLogger } = getState();

setSLogger(logger("combined", schedulerLogFile));
```

Will it be immediately available in sLogger? Of course, linters complain about maybe null etc... I'm ok for the check, but because we don't have a render loop here, I'd need a `delay` or `while` to be implemented if it is not available right away...

**Reply by dbritto-dev** [reply]
@HarikalarKutusu on react you should use `useStore` with zustand vanilla

Here's an example:

```ts
import { useStore } from 'zustand'
import { store } from './store' 

const useBoundStore = (selector) => useStore(store, selector)
```

**Reply by HarikalarKutusu** [reply]
I have no use of React on a microservice. I just want my store to be accessable from workers to create chained jobs etc. So vanilla only...

**Reply by dbritto-dev** [reply]
BTW, if you can create a basic example of your case I can help you with that

**Reply by HarikalarKutusu** [reply]
Actually I posted the basic example above, but I could not pass the problem I mentioned above.
Let me create a github repo thou...

**Reply by dbritto-dev** [reply]
Cool, I'll create an example for you

**Reply by HarikalarKutusu** [reply]
Here is the repo. It cannot be more basic than it is with logging, scheduling, and related workers. But at this stage it is not reaching at that point anyway. Info is in readme file... I'd be very happy to find out what is wrong with this rather simple code...

https://github.com/HarikalarKutusu/zustand-vanilla-bree-test

**Reply by HarikalarKutusu** [reply]
This is the problem: https://github.com/HarikalarKutusu/zustand-vanilla-bree-test/blob/06d93663d9fa2bcc6920517ba7dd8670b56e4945/src/index.ts#L60
